### PR TITLE
Remove the full_dtensor SPMD backend

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #4218
* __->__ #4217

Now that spmd_types is the default backend (thanks to @pianpwk) and partial_dtensor is the fallback, drop `parallelism.spmd_backend == "full_dtensor"` and everything that only existed to support it. This makes people's life easier.

### CODE REVIEW

File: torchtitan/components/loss.py [hunk 1]
```diff
@@ -36,9 +36,6 @@ def cross_entropy_loss(
     global_vocab_size: int | None = None,
 ) -> torch.Tensor:
     """Cross-entropy loss with sum reduction for token-based normalization."""
-    if isinstance(pred, DTensor) and isinstance(labels, DTensor):
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:39]: to confirm: for partial_dtensor with TP, the output are not DTensor?

File: torchtitan/distributed/spmd_types.py [hunk 2]
```diff
@@ -132,6 +134,29 @@ def maybe_set_sparse_mesh() -> Iterator[None]:
         yield
 
 
+def validate_config(parallel_dims: "ParallelDims", model: nn.Module) -> None:
```
Comment (tianyu-l) [inline torchtitan/distributed/spmd_types.py]: this should happen in update_from_config at config time

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/components/loss.py (hunk 1 above)
Comment (fegin) [reply torchtitan/components/loss.py:39]: `pred` is a DTensor; `labels` is not.

**Comment by pianpwk** [comment]
just curious, there's `local_map_*` in models, but I'm not sure if they're removable, or if `partial_dtensor` backend still relies on them

https://github.com/pytorch/torchtitan/blob/ad09b17aee8a201db5f7a93ed8bdc92004302e74/torchtitan/models/qwen3_5/model.py#L343-L357

File: torchtitan/components/loss.py [hunk 3]
```diff
@@ -221,65 +218,6 @@ def backward(  # pyrefly: ignore[bad-override]
...
-            )
-        return _LossParallelCrossEntropy.apply(
-            flat_pred,
-            flat_labels,
-            mesh.get_group("tp"),
-            pred.shape[-1],
-            "sum",
-        )
-
-    return _local_cross_entropy(pred, labels)
```
Comment (pianpwk) [inline torchtitan/components/loss.py:280]: rip

**Comment by fegin** [comment]
> just curious, there's `local_map_*` in models, but I'm not sure if they're removable, or if `partial_dtensor` backend still relies on them
> 
> https://github.com/pytorch/torchtitan/blob/ad09b17aee8a201db5f7a93ed8bdc92004302e74/torchtitan/models/qwen3_5/model.py#L343-L357

Good point, it is unclear. I would proceed to land this one. Without full dtensor, it is easier for Agent to infer.

**Review by pianpwk — APPROVED** [review]
